File: Data Structures & Algorithms/longest-common-subsequence/submission-1.java
class Solution {
    // Time Complexity: O(m x n)
    // Space ComplextiY: O(m x n) - haven't optimize, could be optimize to O(Min(m, n));
    public int longestCommonSubsequence(String text1, String text2) {
        int l1 = text1.length();
        int l2 = text2.length();
        int[][] dp = new int[l1 + 1][l2 + 1]; // space for base case empty string;

        for (int i = 1; i <= l1; i++) {
            for (int j = 1; j <= l2; j++) {
                if (text1.charAt(i - 1) == text2.charAt(j - 1)) {
                    dp[i][j] = dp[i-1][j-1] + 1;
                } else {
                    dp[i][j] = Math.max(dp[i-1][j], dp[i][j-1]);
                }
            }
        }

        return dp[l1][l2];
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.py
class Solution:
    # Time Complexity: O(2n)
    # Space Complexity: O(1)
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0:
            return 0

        start = 0
        end = 0
        set_of_char = set()
        max_len = 0 
        while end < len(s):
            while s[end] in set_of_char:
                set_of_char.remove(s[start])
                start += 1
            max_len = max(end - start + 1, max_len)
            set_of_char.add(s[end])
            end += 1

        return max_len
        
